## 10.1 Sorted Merge

In [6]:
function sortedMerge(A, B) {
    let a = A.length - 1;
    let b = B.length - 1;
    let z = A.length + B.length - 1;
    
    while(a >= 0 && b >= 0 && z >= 0) {
        if(A[a] >= B[b]) {
            A[z] = A[a];
            a--;
        }
        else {
            A[z] = B[b];
            b--;
        }
        z--;
    }
    
    // if we already went through all of B
    // and A is the only one left
    while(a > 0) {
        A[z] = A[a];
        a--;
        z--;  
    }
    
    // if we already went through all of A
    // and B is the only one left
    while(b >= 0) {
        A[z] = B[b];
        b--;
        z--;
    }
    
    return A;
}

var a = [5, 13, 17, 28, 37, 48, 49, 61, 63, 74];
var b = [8, 14, 16, 45, 47, 51, 52, 54, 56, 77];
// [5, 8, 13, 14, 16, 17, 28, 37, 45, 47, 48, 49, 51, 52, 54, 56, 61, 63, 74, 77];
console.log(sortedMerge(a, b));

[
   5,  8, 13, 14, 16, 17, 28,
  37, 45, 47, 48, 49, 51, 52,
  54, 56, 61, 63, 74, 77
]


### Book Solutions:
* exactly the same as my solution
* Solution:
    1. start from the back of array A which has enough space for array B
    2. do a merge by comparing elements from both A and B and add the larger element at the particular spot
    3. if there are only elements from B left, add it to A, else just ignore it since A's elements are already in order at that spot
        - I didn't realize that the elements in A are already ordered so the extra while loop is not needed

In [8]:
function merge(A, B) {
    let a = A.length - 1;
    let b = B.length - 1;
    let z = A.length + B.length - 1;
    
    while(a >= 0 && b >= 0 && z >= 0) {
        if(A[a] >= B[b]) {
            A[z] = A[a];
            a--;
        }
        else {
            A[z] = B[b];
            b--;
        }
        z--;
    }
    
    // removed the while loop that goes through A
    
    // if we already went through all of A
    // and B is the only one left
    while(b >= 0) {
        A[z] = B[b];
        b--;
        z--;
    }
    
    return A;
}

var a = [5, 13, 17, 28, 37, 48, 49, 61, 63, 74];
var b = [8, 14, 16, 45, 47, 51, 52, 54, 56, 77];
// [5, 8, 13, 14, 16, 17, 28, 37, 45, 47, 48, 49, 51, 52, 54, 56, 61, 63, 74, 77];
console.log(merge(a, b));

[
   5,  8, 13, 14, 16, 17, 28,
  37, 45, 47, 48, 49, 51, 52,
  54, 56, 61, 63, 74, 77
]


## 10.2 Group Anagrams

In [40]:
function groupAnagrams(arr) {
    let anagrams = {};
    arr.forEach(str => {
        // sorts the string then trims any white space
        let sorted = str.split('').sort().join('').trim();
        if(anagrams[sorted] === undefined) {
            anagrams[sorted] = [];
        }
        anagrams[sorted].push(str);
    })
    
    let grouped = [];
    
    for(let sorted in anagrams) {
        grouped = grouped.concat(anagrams[sorted]);
    }
    
    return grouped;
}

var arr = ['nose', 'eyes', 'ears', 'seon', 'sear', 'seye', 'random', 'ye es'];
groupAnagrams(arr);

[
  'nose',  'seon',
  'eyes',  'seye',
  'ye es', 'ears',
  'sear',  'random'
]

### Book Solutions:
* words only need to be grouped up by their anagrams; don't need to do any other sorting
* exactly the same as my solution
* Solution:
    1. create a hash table for the anagrams
        - key = sorted anagram
        - value = array full of each word in the arr that matches the sorted anagram
    2. for each word in the array, sort them
    3. if the sorted word is in the hash table, push them into the array
    4. else, add a new entry into the hash table with the sorted word as the key and an empty array as the value
    5. then join all of the entries together to make one big array
* time complexity: O(n * klogk)
    - O(n) b/c we have to make a pass through the entire list to add elements into the hash table and to put them in order into the original array
    - O(klogk) for sorting each string in the array where k = longest string in the array
* space complexity: O(n)
    - this is b/c we are using a hash table to keep track of all the anagrams in the list

## 10.3 Search in Rotated Array:
* doesn't solve edge case where smallest number is at the end of the list

In [44]:
function binarySearch(arr, item) {
    return _binarySearch(item, arr, 0, arr.length - 1);
}

function _binarySearch(item, arr, start, end) {
    let mid = Math.trunc(start + ((end - start) / 2));
    
    if(end < start) {
        return -1;
    }
    if (arr[mid] === item) {
        return mid;
    }
    else if(arr[mid] > item) {
        return _binarySearch(item, arr, start, mid - 1);
    }
    else {
        return _binarySearch(item, arr, mid + 1, end);
    }
}

function rotatedBinarySearch(arr, item) {
    return _rotatedBinarySearch(item, arr, 0, arr.length - 1);
}

function _rotatedBinarySearch(item, arr, start, end) {
    let mid = Math.trunc(start +((end - start) / 2));
    console.log({mid, start, end})
    if(end < start) {
        return -1;
    }
    if(arr[mid] === item) {
        return mid;
    }
    else if(arr[mid] <= item && item <= arr[end]) {
        return _rotatedBinarySearch(item, arr, mid + 1, end);
    }
    else if(arr[start] <= item && item <= arr[mid - 1]) {
        return _rotatedBinarySearch(item, arr, start, mid - 1);
    }
}

var input = [15, 16, 19, 20, 25, 1, 3, 4, 5, 7, 10, 14];
console.log(binarySearch(input, 5));
console.log(rotatedBinarySearch(input, 5));

8
{ mid: 5, start: 0, end: 11 }
{ mid: 8, start: 6, end: 11 }
8


### Book Solutions:
* modification of binary search
* Solution:
    1. if left side is normally ordered
        - check to see if the item is within the range of the left side
        - if it is, check recurse on left side
        - else, recurse on right side
    2. else if right side is normally ordered
        - check to see if the item is within the range of the right side
        - if it is, recurse on right side
        - else, recurse on the left side
    3. else if arr[left] = arr[mid] meaning that the left or right half is all repeated
        - check to see if the right half has any repeats
            - if it doesn't, then recurse on it
        - else if it also has repeats, then check both halves
            - first check the left half.
            - if it is not successful ( === -1 ), then check the right half
            - else, just return the result from the left half
* time complexity: O(log n)
    - this is b/c it is just a modified version of binary search
    - and if all the elements are unique
* worst case scenario: O(n)
    - this happens when there are a bunch of duplicates that result in us having to search for both halves

In [31]:
function search(arr, left, right, item) {
    let mid = Math.trunc( (left + right) / 2);
    // found item
    if(item === arr[mid]) {
        return mid;
    }

    // cannot find item
    if(right < left) {
        return -1;
    }
    
    /* Either the left or right half must be normally ordered.
     * Find out which side is normally ordered, then use the normally ordered
     half to figure out which side to find the item.
    */

    // left half is normally ordered
    if(arr[left] < arr[mid]) {
        // search left if item is between left and mid
        if(item >= arr[left] && item < arr[mid]) {
            return search(arr, left, mid - 1, item);
        }
        // else search right
        else {
            return search(arr, mid + 1, right, item);
        }
    }
    // right half is normally ordered
    else if (arr[mid] < arr[left]) {
        // search right side
        if(item > arr[mid] && item <= arr[right]) {
            return search(arr, mid + 1, right, item);
        }
        // else search left side
        else {
            return search(arr, left, mid - 1, item);
        }
    }
    // left or right half is all repeats
    else if (arr[left] === arr[mid]) {
        // if right side is different, then search it
        if(arr[mid] !== arr[right]) {
            return search(arr, mid + 1, right, item);
        }
        // else we have to search both halves
        else {
            // search left half
            let result = search(arr, left, mid - 1, item);
            // if left half is not successful, search right half
            // else, return the result
            return result === -1 ? search(arr, mid + 1, right, item) : result;
        }
    }
    return -1;
}

var input = [15, 16, 19, 20, 25, 1, 3, 4, 5, 7, 10, 14];
console.log(search(input, 0, input.length - 1, 5));
var input2 = [3, 4, 5, 7, 10, 14, 15, 16, 19, 20, 25, 1];
console.log(search(input2, 0, input2.length - 1, 25))
    

8
10


## 10.4 Sorted Search, No Size

In [61]:
function getEnd(arr, item) {
    let found = false;
    let index = 2;
    let counter = 0;
    while(!found) {
        counter++;
        if(arr[index + index] !== undefined) {
            index += index;
        }
        index += 2;
        if(arr[index] === undefined) {
            index--;
            found = true;
        }
    }
    console.log({counter, index})
    return index;
}

function sortedSearch(arr, item) {
    let end = getEnd(arr, item);
    return end === -1 ? end : _sortedSearch(arr, item, 0, end);
}

function _sortedSearch(arr, item, start, end) {
    let mid = Math.trunc( start + ((end - start) / 2));
    if(end < start) {
        return -1;
    }
    if(arr[mid] === item) {
        return mid;
    }
    else if(arr[mid] > item) {
        return _sortedSearch(arr, item, start, mid - 1);
    }
    else {
        return _sortedSearch(arr, item, mid + 1, end);
    }
}

var listy = [1, 3, 5, 6, 10, 12, 24, 67, 83, 100];
console.log(sortedSearch(listy, 100))
console.log('\n');

var long_listy = [2, 12, 32, 41, 64, 78, 82, 87, 90, 97, 108, 117, 123, 141, 150, 154, 162, 172, 173, 199, 201, 240, 248, 254, 258, 264, 269, 271, 281, 291, 294, 295, 321, 322, 333, 362, 363, 366, 381, 405, 409, 412, 415, 427, 435, 461, 464, 467, 471, 505, 516, 531, 552, 563, 565, 571, 582, 584, 602, 610, 620, 629, 631, 641, 644, 646, 662, 671, 688, 691, 718, 740, 750, 753, 764, 767, 779, 787, 801, 812, 838, 843, 846, 854, 858, 861, 865, 891, 894, 926, 955, 961, 962, 973, 977, 981, 992, 993, 997, 1000];
console.log(`Binary Search result for index: ${binarySearch(long_listy, 787)}`);
console.log('\n');
console.log(sortedSearch(long_listy, 787));

var longer_listy = Array.from( {length: 1000}, (el, i) => i);
console.log(sortedSearch(longer_listy, 777));

var longest_listy = Array.from({ length: 10000}, (el, i) => i);
console.log(sortedSearch(longest_listy, 7777));

{ counter: 3, index: 9 }
9


Binary Search result for index: 77


{ counter: 23, index: 99 }
77
{ counter: 252, index: 999 }
777
{ counter: 916, index: 9999 }
7777


### Book Solutions:
* modification of binary search except we have to find the length ourselves since Listy doesn't have keep track of it
* Solution:
    1. first find the length of the sublist we want to do binary search on
        - to do this, we continually double the index until we either reach a value that is greater than our item or we have reached a point past the length of Listy
    2. then we just do binary search with the length that we found
* time complexity: O(log n)
    - didn't really impact the normal binary search by looking for the length
    - O(log n) for finding the length since we double the index every time
    - and O(log n) for the binary search

In [ ]:
function search(list, value) {
    let index = 1;
    while(index < list.length && list[index] < value) {
        index *= 2;
    }
    return binarySearch(list, value, Math.trunc(index / 2), index);
}

function binarySearch(list, value, low, high) {
    let mid;
    
    while(low <= high) {
        mid = Math.trunc( (low + high) / 2);
        let middle = list[mid];
        if(middle > value || middle === -1) {
            high = mid - 1;
        }
        else if (middle < value) {
            low = mid + 1;
        }
        else {
            return mid;
        }
    }
    return -1;
}

## 10.5 Sparse Search

In [18]:
function binarySearch(arr, item) {
    return _binarySearch(arr, item, 0, arr.length - 1);
}

function _binarySearch(arr, item, start, end) {
    let mid = Math.trunc(start + ((end - start) / 2));
    
    if(end < start) {
        return -1;
    }
    else if (arr[mid] === item) {
        return mid;
    }
    else if (arr[mid] > item) {
        return _binarySearch(arr, item, start, mid - 1);
    }
    else {
        return _binarySearch(arr, item, mid + 1, end);
    }
}

function sparseBinarySearch(arr, item) {
    return _sparseBinarySearch(arr, item, 0, arr.length - 1);
}

function _sparseBinarySearch(arr, item, start, end) {
    let mid = Math.trunc(start + ((end - start) / 2));
    let itemCode = item[0].toLowerCase().charCodeAt(0);
    
    if(end < start) {
        return  -1;
    }
    else if (arr[mid] === item) {
        return mid;
    }
    // if we have an empty string at the mid
    // then check both sides for any items
    else if(arr[mid].length === 0) {
        /*
            1. find the right and left indices that don't have an empty object
            2. if one of them finds something, check if the object @ index is the item
            3. if it isn't then depending on whether it is the left or right will have 2 options
                - if it is left and arr[left][0].charNumber > item[0].char then go from index --> end, else, go from start --> index - 1
                - else if it is right and arr[right][0].charNum < item[0].char, then go from index + 1 --> end, else go from start --> index - 1
        */
        let left = mid - 1;
        let right = mid + 1;
        while(true) {
            if(left < 0 || right > arr.length - 1) {
                break;
            }
            else if(arr[left].length !== 0) {
                if(arr[left] === item) {
                    return left;
                }
                else if(arr[left].toLowerCase().charCodeAt(0) >= itemCode) {
                    end = left;
                    break;
                }
                else {
                    start = left;
                    break;
                }
            }
            else if (arr[right].length !== 0) {
                if(arr[right] === item) {
                    return right;
                }
                else if(arr[right].toLowerCase().charCodeAt(0) <= itemCode) {
                    start = right;
                    break;
                }
                else {
                    end = right;
                    break;
                }
            }
            else {
                left--;
                right++;
            }
        }
        
        if(left < 0 || right > arr.length - 1) {
            return -1;
        }
        return _sparseBinarySearch(arr, item, start, end);
    }
    else if (arr[mid].toLowerCase().charCodeAt(0) >= itemCode) {
        return _sparseBinarySearch(arr, item, start, mid - 1);
    }
    else {
        return _sparseBinarySearch(arr, item, mid + 1, end);
    }
}

var sparse_list = ['at', '', '', '', 'ball', '', '', 'car', '', '', 'dad', '', ''];
console.log(sparseBinarySearch(sparse_list, 'ball')) // 4
console.log(sparseBinarySearch(sparse_list, 'car')) // 7
console.log(sparseBinarySearch(sparse_list, 'dad')) // 10
console.log(sparseBinarySearch(sparse_list, 'at')) // 0

4
7
10
0


### Book Solutions:
* modification of binary search
* pretty much the same as my solution
* Solution:
    1. if the user wants to find an empty string, then return -1
        - really up to the interview whether this should be the case
        - if they don't want this, then just return the index of the first empty string you find
    2. find the middle element of the array
    3. if it is empty, then check for the first non-empty string in the array
        - can do this by looping on the left and right side simultaneously and the first one to have a non-empty string will be used as the new mid
        - so left-- while right++ from the midpoint
        - and we break once left and right have gone past the subarray parameters
    4. then once we have our mid value, we can then do the normal binary search procedure except we compare strings
        - if string === mid, then return mid
        - if string < mid, then search from start --> mid - 1
        - else if string > mid, then search from mid + 1 --> end
* time complexity: O(n)
    - this is b/c for very sparse arrays, you would have to check the entire array for a non-empty string
    - there isn't a way to improve upon this

In [32]:
function _search(strings, str, first, last) {
    if(first > last) {
        return -1;
    }
    
    let mid = Math.trunc((last + first) / 2);
    
    // if mid is empty string, find closest non-empty string
    if(strings[mid].length === 0) {
        let left = mid - 1;
        let right = mid + 1;
        while(true) {
            if(left < first && right > last) {
                return -1;
            }
            else if (right <= last && strings[right].length !== 0) {
                mid = right;
                break;
            }
            else if (left >= first && strings[left].length !== 0) {
                mid = left;
                break;
            }
            right++;
            left--;
        }
    }
    
    // check for string, and recurse if necessary
    if(str === strings[mid]) {
        return mid;
    }
    else if (str > strings[mid]) {
        return _search(strings, str, mid + 1, last);
    }
    else {
        return _search(strings, str, first, mid - 1);
    }
}

function search(strings, str) {
    if(strings == null || str == null || str === "") {
        return -1;
    }
    return search(strings, str, 0, strings.length - 1);
}

## 10.6 Sort Big File

* Assumptions:
    1. memory isn't an issue
    2. this can be done on one machine
    3. assume string length per line is variable
* Algorithm:
    1. break up file into various chunks
        - four 5gb files, or five 4gb files or ten 2gb files, etc
    2. for each file, we want to sort them
        - radix sort for each one from leftmost char in string to rightmost
    3. once each file is sorted, we merge them together
        - file1 + file 2
        - then file3 + file4
        - then (file1 + file2) + (file3 + file4)

### Book Solutions:
* pretty much the same as my solution
* if interview gives you a size limit, this means that they don't want you to put all the elements into memory
* thus we should only bring part of the data into memory
* we essentially split the file into chunks and sort them each individually
* then once all the chunks are sorted, we can then merge them together
* this is known as external sort

## 10.7 Missing Int:
* Info:
    - 4 billion integers
    - each integer is non-negative so >= 0
    - assume each int is 4 bytes
    - there's 1GB of memory to do the task
    - if using memory, can only process 250 million integers at a time
* Algorithm:
    1. make 1 pass through the input file and keep track of the max number in the list
    2. return max + 1 since it will not be in the file
* O(n) time complexity where n = number of integers and O(1) space complexity since only needs to keep track of the max

### Book Solutions:
* there are 2$^{32}$ (4 billion) possible integers and 2$^{31}$ non-negative integers so there are some duplicates
* 1Gb of memory = 8 billion bits
    - 1 gigabyte = 1 billion bytes
    - 1 byte = 8 bits
    - therefore 1Gb = 8 billion bits
* since there are 8 billion bits, we can map all integers to a distinct bit with all the available memory since there are 4 billion possible integers
* Solution:
    1. create a bit vector(BV) with 4 billion bits
    2. initialize bit BV with all 0s
    3. scan all numbers (num) from the file and call BV.set(num, 1)
    4. now scan BV again from the 0th index
    5. return the first index which has a value 0
* Follow-Up Solution:
    1. divide integers into blocks of some size, e.g. 1000
    2. we know that each block has a certain range of integers, for example block 0 = 0 - 999
    3. then if we find a block that has duplicates, then we use the bit vector approach from above to find the missing int.

## 10.8 Find Duplicates

In [13]:
// finds duplicates by modifying insertion Sort
// while insertion sort operates, we can also find duplicates too
// bc we essentially check a new value with an already sorted sublist
// of values
// total time: O(n^2), total space = O(1)

function insertionSort(arr) {
    for(let i = 1; i < arr.length; i++) {
        let currentValue = arr[i];
        let position = i;
        
        while(position > 0 && arr[position - 1] > currentValue) {
            arr[position] = arr[position - 1];
            position--;
        }
        
        if(currentValue === arr[position - 1]) {
            console.log(currentValue);
        }
        
        arr[position] = currentValue;
    }
}

var alist = [54, 26, 93, 17, 77, 77, 77, 77, 31, 44, 55, 20];
insertionSort(alist);

77
77
77


### Book Solutions
* 4 kilobytes of memory so we can address up to 8 * 4 * 2$^{10}$ bits which is > 32000
* so knowing this, we can create a bit vector with 32000 bits and each bit represents one integer
* Solution:
    1. create the bit vector with length of 32000
    2. pass through the array
        - if the bit[arr_value] = 0, then set it
        - if the bit[arr_value] = 1, then print it since it is a duplicate

## 10.9 Sorted Matrix Search

In [50]:
function findMid(start, end) {
    return Math.trunc(start + ((end - start) / 2));
}

function matrixBinarySearch(matrix, item) {
    return _matrixBinarySearch(matrix, item, 0, matrix.length - 1, 0, matrix[0].length - 1);
}

function _matrixBinarySearch(matrix, item, startRow, endRow, startCol, endCol) {
    let midRow = findMid(startRow, endRow);
    let midCol = findMid(startCol, endCol);
    if(endRow < startRow || endCol < startCol) {
        return -1;
    }
    if(matrix[midRow][midCol] === item) {
        return {midRow, midCol};
    }
    // check top left quadrant
    // if item is less than middle
    // check from startRow --> midRow
    // and startCol --> midCol
    else if (matrix[midRow][midCol] > item) {
        if(matrix[midRow][0] > item) {
            midRow--;
        }
        else if (matrix){
            midCol--;  
        }
        return _matrixBinarySearch(matrix, item, startRow, midRow, startCol, midCol);
    }
    // check bottom right quadrant
    // if item is greater than middle
    // check from midRow --> endRow
    // check from midCol --> endCol
    else {
        if(matrix[midRow][matrix[0].length - 1] < item) {
            midRow++;
        }
        else {
            midCol++; 
        }
        return _matrixBinarySearch(matrix, item, midRow, endRow, midCol, endCol);
    }
}

var matrix = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8]
];

matrixBinarySearch(matrix, 6);

-1

In [121]:
// O(r + c)

function matrixSearch(matrix, item) {
    let result = -1;
    for(let r = 0; r < matrix.length; r++) {
        let c = matrix[0].length - 1;
        while(c >= 0) {
            if(matrix[r][c] === item){
                result = {r, c};
                return result;
            }
            c--;
        }
    }
    return result;
}

var matrix = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8]
];

console.log(matrixSearch(matrix, 7)); // r: 2, c: 1
console.log(matrixSearch(matrix, 9)); // -1

{ r: 2, c: 1 }
-1


### Book Solutions:
* we know that:
    - we can split the matrix into 4 quadrants
        - top-left
        - top-right
        - bottom-left
        - bottom-right
    - if we have a value called mid, then:
        - all values in the top-left quadrant, including itself, will have a value <= to it
        - and all values in the bottom-right quadrant, including itself, will have a value >= to it
* Solution:
    1. find the mid of the matrix by using row and col length
    2. then the mid will be used to find the exact row and col index
        - row = Math.floor(mid / col_length)
        - col = mid % col_length
    3. then check matrix[row][col] for our value
        - if it is equal, return {row, col}
        - if value < matrix[row][col], then recurse on start --> mid
        - else if value > matrix[row][col], then recurse on mid + 1 --> end

In [46]:
function sortedMatrixSearch(matrix, value, front, back) {
    if(matrix === undefined) {
        return 'No matrix found';
    }
    
    let m = matrix.length;
    let n = matrix[0].length;
    
    if(front === undefined && back === undefined) {
        front = 0;
        back = m * n;
    }
    
    console.log({front, back});
    
    if(front > back) { return - 1; }
    
    let mid = Math.trunc( (front + back) / 2);
    let row = Math.trunc( mid / n);
    let col = mid % n;
    
    console.log({row, col});
    
    if(matrix[row][col] === value) {
        return {row, col};
    }
    else if (value < matrix[row][col]) {
        return sortedMatrixSearch(matrix, value, front, mid);
    }
    else {
        return sortedMatrixSearch(matrix, value, mid + 1, back);
    }
}


var matrix = [
    [0, 1, 2],
    [3, 4, 5],
    [6, 7, 8]
];

sortedMatrixSearch(matrix, 6);

{ front: 0, back: 9 }
{ row: 1, col: 1 }
{ front: 5, back: 9 }
{ row: 2, col: 1 }
{ front: 5, back: 7 }
{ row: 2, col: 0 }


{ row: 2, col: 0 }

## 10.10 Rank from Stream

In [36]:
// modification of Selection Sort to keep track of items
// O(n^2) insertion

class Stream {
    constructor(stream) {
        this.stream = stream;
        this.rank = {};
        this.tracker = [];
    }
    
    getStream() {
        this.stream.forEach(int => {
            this.methodTrack(int);
        })
    }
    
    methodTrack(x) {
        this.tracker.push(x);
        if(this.rank[x] === undefined) {
            this.rank[x] = 0;
        }
        if(this.tracker.length === 1) {
            return;
        }
        let position = this.tracker.length - 1;
        while(position > 0 && this.tracker[position - 1] > x) {
            if(this.tracker[position] !== this.tracker[position - 1]) {
                this.rank[this.tracker[position - 1]]++;
            }
            this.tracker[position] = this.tracker[position - 1];
            position--;
        }
        
        this.tracker[position] = x;
        this.rank[x] = (position - 1) > - 1 ? this.rank[this.tracker[position - 1]] + 1 : 0;
        return;
    }
    
    getRankOfNumber(x) {
        return this.rank[x];
    }
    
    getRanks() {
        return this.rank;
    }
}

var stream = [5, 1, 4, 4, 5, 9, 7, 13, 3];
var S = new Stream(stream);
S.getStream();
console.log(S.getRanks());
console.log(S.getRankOfNumber(1));
console.log(S.getRankOfNumber(3));
console.log(S.getRankOfNumber(4));

{ '1': 0, '3': 1, '4': 3, '5': 5, '7': 6, '9': 7, '13': 8 }
0
1
3


In [11]:
// using a binary search tree
// algorithm should be O(log n)

class TreeNode {
    constructor(key, val, left = null, right = null, parent = null, duplicates = 0) {
        this.key = key;
        this.payload = val;
        this.left = left;
        this.right = right;
        this.parent = parent;
        this.duplicates = duplicates;
    }
    
    hasLeftChild() {
        return this.left !== null;
    }
    
    hasRightChild() {
        return this.right !== null;
    }
    
    isLeftChild() {
        return this.parent !== null && this.parent.left === this;
    }
    
    isRightChild() {
        return this.parent !== null && this.parent.right === this;
    }
    
    isRoot() {
        return this.parent === null;
    }
    
    isLeaf() {
        return !(this.hasAnyChildren() );
    }
    
    hasAnyChildren() {
        return this.hasLeftChild() || this.hasRightChild();
    }
    
    hasBothChildren() {
        return this.hasLeftChild() && this.hasRightChild();
    }
    
    replaceNodeData(key, value, left, right) {
        this.key = key;
        this.payload = value;
        this.left = left;
        this.right = right;
        if(this.hasLeftChild()) {
            this.left.parent = this;
        }
        if(this.hasRightChild()) {
            this.right.parent = this;
        }
    }
    
    findSuccessor() {
        let successor;
        if(this.hasRightChild()) {
            successor = this.right.findMin();
        }
        else {
            if(this.parent !== null) {
                if(this.isLeftChild()) {
                    successor = this.parent;
                }
                else {
                    this.parent.right = null;
                    successor = this.parent.findSucessor();
                    this.parent.right = this;
                }
            }
        }
        return successor;
    }
    
    findMin() {
        let current = this;
        while(current.hasLeftChild()) {
            current = current.left;
        }
        return current;
    }
    
    spliceOut() {
        if(this.isLeaf()) {
            if(this.isLeftChild()) {
                this.parent.left = null;
            }
            else {
                this.parent.right = null;
            }
        }
        else if (this.hasAnyChild()) {
            if(this.hasLeftChild()) {
                if(this.isLeftChild()) {
                    this.parent.left = this.left;
                }
                else {
                    this.parent.right = this.left;
                }
                this.left.parent = this.parent;
            }
            else {
                if(this.isleftChild()) {
                    this.parent.left = this.right;
                }
                else {
                    this.parent.right = this.right;
                }
                this.right.parent = this.parent;
            }
        }
    }
}


const root = Symbol('root');

class BinarySearchTree {
    constructor() {
        this[root] = null;
        this.size = 0;
    }
    
    get length() {
        return this.size;
    }
    
    inOrder(node = this[root]) {
        if(node !== null) {
            this.inOrder(node.left);
            console.log([node.key, node.payload]);
            this.inOrder(node.right);
        }
    }
    
    incrementRightSubtree(node) {
        if(node !== null) {
            this.incrementRightSubtree(node.left);
            node.payload++;
            this.incrementRightSubtree(node.right);
        }
    }
    
    put(key, val) {
        if(this[root] !== null) {
            this._put(key, val, this[root]);
        }
        else {
            this[root] = new TreeNode(key, val);
        }
        this.size++;
    }
    
    _put(key, val, currentNode) {
        if (key < currentNode.key) {
            // if key is less than currentNode then increment its payload
            // and also increment every item in its right subtree b/c there is 1 more
            // item that they are greater than
            currentNode.payload++;
            if(currentNode.hasRightChild()) {
                this.incrementRightSubtree(currentNode.right);
            }
            if(currentNode.hasLeftChild()) {
                this._put(key, val, currentNode.left);
            }
            else {
                // if current key has found a place, then its value will be the current node's value - 1 - # of duplicates
                // for example: 5,1,4,4,3
                // in this case, since there are two 4s, its value would be 2 instead of 1
                // and when we add 3, we see that 3 is (3 - 1 - 1) b/c there are two 4s there
                // rank[3] would be 1 which is correct b/c the number 1 is the only one that 3 is larger than
                // rather than 2 if we did not count the duplicate in, i.e. 3 - 1
                currentNode.left = new TreeNode(key, currentNode.payload - 1 - currentNode.duplicates, null, null, currentNode);
            }
        }
        else if (key === currentNode.key) {
            // if we get the same key, just increment it since the problem said rank = # of items <= current item
            // and we increment duplicates too so that when we add a right child, it will not count the duplicate
            // for its ranking
            currentNode.payload++;
            currentNode.duplicates++;
        }
        else {
            if(currentNode.hasRightChild()) {
               this._put(key, val, currentNode.right);
            }
            else {
                currentNode.right = new TreeNode(key, currentNode.payload + 1, null, null, currentNode);
            }
        }
    }
    
    
    get(key) {
        if(this[root] === null) {
            return undefined;
        }
        const result = this._get(key, this[root]);
        return result ? result.payload : undefined;
    }
    
    _get(key, currentNode) {
        if(currentNode === null) {
            return undefined;
        }
        else if(currentNode.key === key) {
            return currentNode;
        }
        else if (key < currentNode.key) {
            return this._get(key, currentNode.left);
        }
        else {
            return this._get(key, currentNode.right);
        }
    }
}

class Stream {
    constructor(stream) {
        this.stream = stream;
        this.tracker = new BinarySearchTree();
    }
    
    methodTracker() {
        this.stream.forEach(int => {
            this.tracker.put(int, 0);
        })
    }
    
    getRankOfNumber(x) {
        return this.tracker.get(x);
    }
    
    getTracker() {
        this.tracker.inOrder();
    }
}

var stream = [5, 1, 4, 4, 5, 9, 7, 13, 3];
var S = new Stream(stream);
S.methodTracker();
S.getTracker();
console.log(S.getRankOfNumber(1));
console.log(S.getRankOfNumber(3));
console.log(S.getRankOfNumber(4));

[ 1, 0 ]
[ 3, 1 ]
[ 4, 3 ]
[ 5, 5 ]
[ 7, 6 ]
[ 9, 7 ]
[ 13, 8 ]
0
1
3


### Book Solutions:
* the exact same solution as my bst implementation
* time complexity: O(log n)
    - this is for insertion and for updating the ranks if the number goes through the left subtree and you have numbers on the right
    - O(log n) for finding a number's ranking as well
* worst case scenario: O(n)
    - this is when the bst is not balanced and you would essentially have to traverse through all the elements to find a right position or to find the element

## 10.11 Peaks and Valleys

In [3]:
// inspired by bubble sort
// is about O(n^2)

function peaksAndValleys(arr) {
    let passes = arr.length;
    let sorted = arr;
    let sequenced = 0;
    while(passes > 0) {
        sequenced = 0;
        for(let i = 0; i < sorted.length; i++) {
            let currentValue = sorted[i];
            if(i === 0) {
                if(sorted[1] > currentValue) {
                    [sorted[0], sorted[1]] = [sorted[1], sorted[0]];
                }
                else {
                    sequenced++;
                }
            }
            
            // if peak
            if(i % 2 === 0) {
                if(i !== sorted.length - 1) {
                    if(sorted[i - 1] > currentValue || currentValue < sorted[i + 1]) {
                        [sorted[i], sorted[i + 1]] = [sorted[i + 1], sorted[i]];
                    }
                    else {
                        sequenced++;
                    }
                }
                else {
                    if(sorted[i - 1] > currentValue) {
                        [sorted[i], sorted[i - 1]] = [sorted[i - 1], sorted[i]];
                    }
                    else {
                        sequenced++;
                    }
                }
            }
            // if valley
            else {
                if(i !== sorted.length - 1) {
                    if(sorted[i - 1] < currentValue || currentValue > sorted[i + 1]) {
                        [sorted[i], sorted[i + 1]] = [sorted[i + 1], sorted[i]];
                    }
                    else {
                        sequenced++;
                    }
                }
                else {
                    if(sorted[i - 1] < currentValue) {
                        [sorted[i], sorted[i - 1]] = [sorted[i - 1], sorted[i]];
                    }
                    else {
                        sequenced++;
                    }
                }
            }
            passes--;
            if(sequenced === sorted.length) {
                return {sorted, passes};
            }
        }
    }
    return {sorted, passes};
}

var list = [5, 3, 1, 2, 3];
console.log(peaksAndValleys(list));

var list2 = [0, 1, 2, 3, 5, 7, 9];
console.log(peaksAndValleys(list2));

var list3 = [1, 0, 3, 2, 7, 5, 9];
console.log(peaksAndValleys(list3));

var list4 = [5, 1, 3, 2, 3];
console.log(peaksAndValleys(list4))

var long_list = [688, 467, 229, 572, 586, 289, 208, 947, 543, 953, 193, 144, 293, 348, 510, 401, 658, 577, 356, 845, 948, 88, 137, 935, 78, 530, 945, 340, 967, 711, 337, 732, 684, 455, 817, 731, 77, 493, 712, 918, 587, 86, 183, 136, 884, 264, 962, 450, 60, 995];
console.log(peaksAndValleys(long_list));

{ sorted: [ 5, 1, 3, 2, 3 ], passes: 0 }
{ sorted: [
    1, 0, 3, 2,
    7, 5, 9
  ], passes: 0 }
{ sorted: [
    1, 0, 3, 2,
    7, 5, 9
  ], passes: 1 }
{ sorted: [ 5, 1, 3, 2, 3 ], passes: 1 }
{
  sorted: [
    688, 229, 572, 467, 586, 208, 947, 289, 953,
    193, 543, 144, 348, 293, 510, 401, 658, 356,
    845, 577, 948,  88, 935,  78, 530, 137, 945,
    340, 967, 337, 732, 684, 711, 455, 817,  77,
    731, 493, 918, 587, 712,  86, 183, 136, 884,
    264, 962,  60, 995, 450
  ],
  passes: 0
}


### Book Solutions:
* Solution:
    1. iterate through the array and jump by 2 indices starting at index 1
    2. for every iteration, you want to compare i, i - 1, and i + 1
    3. we then take the index with the largest value and put it in the center for a peak
        - this is why we skip by 2 indices b/c if we do this, then the i + 1 value will always be a valley since its neighbors will have a peak
        - if the largest value is i, then don't make swaps
        - if it isn't i, meaning that the largest is one of its neighbors, then swap with i
* time complexity: O(n)
    - we just iterate through the array and make swaps

In [52]:
function maxIndex(arr, a, b, c) {
    let len = arr.length;
    let aValue = a >= 0 && a < len ? arr[a] : Number.MIN_VALUE;
    let bValue = b >= 0 && b < len ? arr[b] : Number.MIN_VALUE;
    let cValue = c >= 0 && c < len ? arr[c] : Number.MIN_VALUE;
    
    let max = Math.max(aValue, bValue, cValue);
    if(aValue === max) {
        return a;
    }
    else if (bValue === max) {
        return b;
    }
    else {
        return c;
    }
}


function sortValleyPeak(arr) {
    for(let i = 1; i < arr.length; i += 2) {
        let biggestIndex = maxIndex(arr, i - 1, i, i + 1);
        if(i !== biggestIndex) {
            [arr[i], arr[biggestIndex]] = [arr[biggestIndex], arr[i]];
        }
    }
    return arr;
}

var long_list = [688, 467, 229, 572, 586, 289, 208, 947, 543, 953, 193, 144, 293, 348, 510, 401, 658, 577, 356, 845, 948, 88, 137, 935, 78, 530, 945, 340, 967, 711, 337, 732, 684, 455, 817, 731, 77, 493, 712, 918, 587, 86, 183, 136, 884, 264, 962, 450, 60, 995];
console.log(sortValleyPeak(long_list));

[
  467, 688, 229, 586, 289, 572, 208, 947, 543,
  953, 193, 293, 144, 510, 348, 658, 401, 577,
  356, 948,  88, 845, 137, 935,  78, 945, 530,
  967, 340, 711, 337, 732, 684, 817, 455, 731,
   77, 712, 493, 918,  86, 587, 183, 884, 136,
  962, 264, 450,  60, 995
]
